File: 9. Intervals/BBG. Largest Overlap of Intervals/Solution.cpp
#include <vector>

struct Interval
{
    int start;
    int end;
    Interval(int start, int end) : start(start), end(end) {}
};
    
int largest_overlap_of_intervals(std::vector<Interval> intervals)
{
    std::vector<std::pair<int, char>> events;

    for (Interval& interval : intervals)
    {
        events.push_back(std::make_pair(interval.start, 's'));
        events.push_back(std::make_pair(interval.end, 'e'));
    }
    std::sort(events.begin(), events.end());

    int largestOverlap {0};
    int currOverlap {0};

    for (auto& [_, e] : events)
        if (e == 's')
        {
            ++currOverlap;
        }
        else
        {
            --currOverlap;
        }
        
        largestOverlap = std::max(largestOverlap, currOverlap);

    return largestOverlap;
}

File: 9. Intervals/BBG. Largest Overlap of Intervals/Solution.py
from typing import List

class Interval:
    def __init__(self, start: int, end: int):
        self.start = start
        self.end = end


def largest_overlap_of_intervals(intervals: List[Interval]) -> int:
    sweeping_line = []

    for interval in intervals:
        sweeping_line.append((interval.start, 's'))
        sweeping_line.append((interval.end, 'e'))

    sweeping_line.sort()

    largest_overlap = 0
    curr_overlap = 0

    for _, char in sweeping_line:
        if char == 's':
            curr_overlap += 1
        else:
            curr_overlap -= 1
        
        largest_overlap = max(largest_overlap, curr_overlap)

    return largest_overlap